ssh_tunnel.go

```diff
@@ -90,14 +91,29 @@ func (tunnel *SSHTunnel) Start() error {
 }
 
 func (tunnel *SSHTunnel) forward(localConn net.Conn) {
-	serverConn, err := ssh.Dial("tcp", tunnel.Server.String(), tunnel.Config)
-	if err != nil {
-		tunnel.logf("server dial error: %s", err)
-		return
+	var (
+		serverConn   *ssh.Client
+		err          error
+		attemptsLeft int = tunnel.MaxConnectionAttempts
+	)
+
+	for {
+		serverConn, err = ssh.Dial("tcp", tunnel.Server.String(), tunnel.Config)
+		if err != nil {
+			attemptsLeft--
+
+			if attemptsLeft <= 0 {
+				tunnel.logf("server dial error: %w: exceeded %d attempts", err, tunnel.MaxConnectionAttempts)
```

Comment:
This will generate a misleading error if the `tunnel.MaxConnectionAttempts` is not set (ie. 0). You have two ways around this:

1. Update `NewSSHTunnel` to set `MaxConnectionAttempts` to 1.
2. Make sure the value displayed is not less than 1.

Personally, I would choose the first option but it's up to you.

Reply:
That's why I was initially setting the attempts to 1 if MaxConnctionAttempts was 0. I should have actually ensured that MaxConnectionAttempts was at least 1. I'm thinking it might make more sense to do that verification in Start(), since the MaxConnectionAttempts could be changed at any point between the NewSSHTunnel() and Start() calls. And that logic/check should be done once instead of in forward() where it would happen for every connection made through the tunnel. If it was done in NewSSHTunnel(), the value could potentially be changed to 0 by the programmer before Start().

What do you think?

Reply:
I don't really mind how you handle it internally as long as the existing behavior is respected and the error message is correct. That's why I would go with #1 as it's the simplest way to cover those two requirements.